### PR TITLE
test(auth): formal behavioral test coverage for all OAuth bucket failover scenarios (Fixes #1783)

### LINKED ISSUE #1783: Task: Formal specification and behavioral test coverage for all OAuth bucket failover scenarios
Parent issue: #1777

## Problem

The OAuth bucket failover system has grown organically across many issues (#1029, #1151, #1159, #1317, #1378, #1598, #1616, #1680) but lacks a unified formal specification of all scenarios and their expected behaviors. Existing tests cover individual bug fixes but do not systematically verify the complete state space. Several critical paths have no behavioral test coverage.

## Goal

Create a formal specification of every bucket failover scenario and write behavioral tests (no mock theater) that verify each one.

### Single-Bucket (Unbucketed) Profile Scenarios

| ID | Scenario | Given | When | Then |
|----|----------|-------|------|------|
| SB-01 | Valid token | Single bucket, token valid | getToken called | Returns access_token, schedules proactive renewal |
| SB-02 | Expired token, refresh succeeds | Token expired, valid refresh_token | getToken called | Acquires refresh lock, refreshes, saves, returns new token |
| SB-03 | Expired token, refresh fails | Token expired, refresh returns null | getToken called | Falls through to disk check, then to auth flow |
| SB-04 | Expired token, no refresh_token | Token expired, empty refresh_token | getToken called | Skips refresh, falls through to auth flow |
| SB-05 | Token revoked server-side (401) | Token valid locally, API returns 401 | RetryOrchestrator retries | First retry allows refresh, second 401 triggers error |
| SB-06 | Token revoked server-side (403) | Token valid locally, API returns 403 | RetryOrchestrator retries | Same as SB-05 |
| SB-07 | Cross-process refresh | Two processes, same expired token | Both call getOAuthToken | Only one refreshes via lock; other reads refreshed token from disk |
| SB-08 | Proactive renewal fires | Token near expiry (within 10% of lifetime) | Timer fires | Acquires lock, refreshes, saves, reschedules |
| SB-09 | Proactive renewal - lock contention | Two processes, both timers fire | Both attempt renewal | One refreshes; other detects changed token and reschedules |
| SB-10 | Auth flow mid-turn timeout | Token dead, auth flow starts mid-turn | Timeout exceeded | Turn fails with auth error, next prompt triggers relogin |

### Multi-Bucket Profile Scenarios

| ID | Scenario | Given | When | Then |
|----|----------|-------|------|------|
| MB-01 | Primary bucket valid | 3 buckets, bucket1 valid | getToken called | Returns bucket1 token |
| MB-02 | Primary 429, failover to bucket2 | bucket1 rate-limited | API returns 429 | BucketFailoverHandler Pass 2 switches to bucket2 |
| MB-03 | Primary 401, failover | bucket1 token revoked | API returns 401 | First 401 retries (refresh attempt), second 401 triggers Pass 1 classify + Pass 2 failover |
| MB-04 | All buckets 429 | All buckets rate-limited | Sequential 429s | All buckets exhausted error with correct `bucketFailureReasons` map |
| MB-05 | All buckets expired, refresh succeeds | All tokens expired, all have refresh_tokens | failover triggered | Pass 2 refreshes first available bucket and switches |
| MB-06 | All buckets expired, refresh fails | All tokens expired, refresh fails for all | failover triggered | Pass 3 foreground reauth attempted for first candidate |
| MB-07 | Foreground reauth succeeds | Pass 3 candidate bucket | authenticate called | Token saved, bucket switched, returns true |
| MB-08 | Foreground reauth fails | Pass 3 candidate bucket | authenticate throws | Classified as reauth-failed, returns false |
| MB-09 | Bucket removed mid-session | 3 buckets, bucket2 removed from store | failover tries bucket2 | Returns no-token, skips to bucket3 |
| MB-10 | Session bucket override | Session overridden to bucket2 | getOAuthToken called | Uses bucket2 instead of bucket1 |
| MB-11 | Peek other buckets (issue1616) | Primary returns null, bucket2 valid | getToken called | Peeks bucket2, switches session, returns bucket2 token |
| MB-12 | ensureBucketsAuthenticated at turn boundary | Multiple unauthenticated buckets | Turn boundary reached | All buckets authenticated eagerly |
| MB-13 | Concurrent failover coalescing | Two concurrent failovers for same bucket | Both call tryFailover | Second awaits first via foregroundReauthInFlightByBucket |
| MB-14 | Cross-process refresh per-bucket | Two processes, bucket1 expired | Both refresh bucket1 | Lock prevents concurrent refresh; non-winner reads disk |

### Proactive Renewal Scenarios

| ID | Scenario | Given | When | Then |
|----|----------|-------|------|------|
| PR-01 | Schedule renewal for valid token | Token with >5min remaining, has refresh_token | scheduleProactiveRenewal | Timer set for (expiry - lead - jitter) |
| PR-02 | Skip renewal for short-lived token | Token with <5min remaining | scheduleProactiveRenewal | No timer set, existing timer cleared |
| PR-03 | Skip renewal for no refresh_token | Token without refresh_token | scheduleProactiveRenewal | No timer set |
| PR-04 | Renewal succeeds | Timer fires, lock acquired | runProactiveRenewal | Token refreshed, saved, new timer scheduled |
| PR-05 | Renewal fails, retry with backoff | Timer fires, refresh returns null | runProactiveRenewal | Failure count incremented, retry scheduled with exponential backoff |
| PR-06 | Renewal stops after MAX failures | 3 consecutive failures | runProactiveRenewal | Timer cleared, no more retries |
| PR-07 | External refresh detected | Another process refreshed | runProactiveRenewal | Detects changed token, reschedules without refreshing |
| PR-08 | Proxy mode skipped | LLXPRT_CREDENTIAL_SOCKET set | scheduleProactiveRenewal | Returns immediately (host process handles refresh) |
| PR-09 | Profile reconfiguration | Profile changed to different buckets | configureProactiveRenewalsForProfile | Old timers cleared, new timers set for new buckets |
| PR-10 | Dedup: same expiry already scheduled | Same token.expiry already has timer | scheduleProactiveRenewal | No-op, existing timer kept |

### Bridge and UI Integration Scenarios

| ID | Scenario | Given | When | Then |
|----|----------|-------|------|------|
| BR-01 | Bridge available at module load | global-oauth-ui.ts imported | Module loads | global.__oauth_add_item is a function |
| BR-02 | Event before UI mount | OAuth URL event fires | Before AppContainer mounts | Event buffered in GlobalOAuthUI |
| BR-03 | Buffer flush on handler attach | Buffered events exist | setAddItem called | All buffered events flushed in order |
| BR-04 | Event during cleanup/remount | Handler detached via clearAddItem | OAuth URL event fires | Event buffered, delivered on next setAddItem |
| BR-05 | Direct delivery when mounted | Handler attached | OAuth URL event fires | Event delivered immediately, no buffering |

---

### User Entry Points / Lifecycle Triggers *(NEW — missing from original catalog)*

The original catalog covers internal mechanism scenarios well but omits the user-perspective entry points that trigger these flows. The same underlying mechanism (e.g., token refresh) behaves differently depending on how it's entered.

| ID | Scenario | Given | When | Then |
|----|----------|-------|------|------|
| UE-01 | `--profile-load` at CLI startup | Multi-bucket profile, no cached tokens | CLI starts with `--profile-load mybucketed` | `configureProactiveRenewalsForProfile` called, `ensureBucketsAuthenticated` triggers eager auth for all buckets before first prompt |
| UE-02 | `/profile load` mid-session | Active session on profile-A | User types `/profile load profile-B` (bucketed) | Old proactive renewal timers cleared, new handler wired, new buckets eagerly authenticated |
| UE-03 | New turn begins (user types prompt) | Session active, tokens valid | User presses Enter on new prompt | `resetSession()` clears `triedBuckets`, `ensureBucketsAuthenticated()` runs eagerly (useGeminiStream.ts L73), then API call proceeds |
| UE-04 | Token expires mid-agent-turn | Agent executing tool loop, token near expiry | Proactive renewal fires while agent is mid-tool-call | Renewal runs in background without interrupting agent's turn; next API call uses refreshed token |
| UE-05 | 429 during agent's tool loop | Agent mid-turn calling APIs, bucket-a rate-limited | API returns 429 during tool execution | RetryOrchestrator triggers `tryFailover`, transparent switch to bucket-b, agent continues unaware |
| UE-06 | User declines auth prompt | OAuth dialog shown to user | User clicks Cancel / says No | Turn fails gracefully, next prompt allows retry; already-authenticated buckets remain usable |
| UE-07 | All tokens die during agent turn | 3 buckets, all expired, all refresh_tokens consumed | Agent mid-turn, next API call | Sequential failover through all buckets, all fail refresh, Pass 3 foreground reauth triggers UI dialog |
| UE-08 | Profile switch clears old state | Active session on bucketed profile-A | Switch to profile-B | Old failover handler, session buckets, proactive timers all cleaned up; new ones created for profile-B |

### Subagent Isolation Scenarios *(NEW — referenced by PR #1720 but absent from catalog)*

PR #1720 (isolate subagent auth state, fixes #1718/#1719) introduced metadata-scoped session buckets. This is one of the more complex parts of the auth system and has no dedicated scenario coverage.

| ID | Scenario | Given | When | Then |
|----|----------|-------|------|------|
| SA-01 | Subagent gets scoped session bucket | Parent on bucket-a, subagent profile also bucketed | Subagent starts | Subagent gets metadata-scoped session bucket independent of parent |
| SA-02 | Concurrent subagents same provider | Two subagents, both using anthropic with different `OAuthTokenRequestMetadata` | Both call `getOAuthToken` | Each gets its own session bucket scope; no cross-contamination |
| SA-03 | Subagent failover doesn't affect parent | Subagent on bucket-a, fails over to bucket-b | Parent makes next API call | Parent still uses its own session bucket (bucket-a), unaffected |
| SA-04 | Subagent eager auth with metadata | Subagent profile has 3 buckets | `authenticateMultipleBuckets` called with metadata | Handler wired with metadata scope, session bucket scoped to subagent |

### Error & Edge Case Scenarios *(NEW)*

| ID | Scenario | Given | When | Then |
|----|----------|-------|------|------|
| EC-01 | Network error during refresh (throws, not null) | Token expired, refresh_token valid | `refreshToken()` throws `Error` (network/timeout) | Error caught, bucket classified as `expired-refresh-failed`, falls through to next recovery step |
| EC-02 | Token store lock stale detection | Lock file exists but stale (>30s old) | `acquireRefreshLock` called with `staleMs:30000` | Stale lock detected and overridden after `staleMs` |
| EC-03 | Auth flow interrupted by SIGINT | OAuth browser flow in progress | User presses Ctrl+C | Graceful cleanup, lock released, no orphaned state |
| EC-04 | Logout then re-auth same session | User has authenticated tokens | User runs `/auth logout` then sends prompt | All tokens cleared, next prompt triggers fresh auth flow |

### Retry Orchestrator Integration Scenarios *(NEW — multi-bucket RO paths missing from catalog)*

The original catalog covers 401/403 only in the single-bucket context (SB-05/SB-06). The full multi-bucket RetryOrchestrator integration is absent.

| ID | Scenario | Given | When | Then |
|----|----------|-------|------|------|
| RO-01 | 401 with bucket failover handler wired | Multi-bucket profile, bucket-a active | API returns 401 | First retry: `tokenRefreshCallback` refreshes. Second 401: triggers `tryFailover` to bucket-b |
| RO-02 | 429 with Retry-After header | Multi-bucket profile | API returns 429 with `Retry-After: 60` | `BucketFailoverHandler` invoked, switches bucket; `Retry-After` respected for original bucket |
| RO-03 | `resetSession` on successful request | Multi-bucket, previously failed over to bucket-b | Next request succeeds | `resetSession()` called by RetryOrchestrator, `triedBuckets` cleared for next request |

---

### Gold-Standard Behavioral Tests (use real implementations)
- **`BucketFailoverHandlerImpl.spec.ts`** — real OAuthManager + MemoryTokenStore; covers MB-01–MB-13 well
- **`global-oauth-ui.test.ts`** — real globalOAuthUI singleton, zero mocks; covers BR-01–BR-05 + edge cases
- **`proactive-renewal-manager.spec.ts`** — real ProactiveRenewalManager + vi.useFakeTimers(); covers PR-01–PR-10
- **`multi-bucket-auth.spec.ts`** — real MultiBucketAuthenticator with callback injection
- **`RetryOrchestrator.test.ts`** — real RetryOrchestrator; covers SB-05/SB-06 and 429 retry logic

### Mixed Coverage (real classes, but vi.fn() stores)
- **`token-access-coordinator.spec.ts`** — lock parameter assertions are structural but valuable
- **`oauth-manager.concurrency.spec.ts`** — real lock contention behavior, inline mock stores
- **`proactive-renewal-cross-process.spec.ts`** — conceptual cross-process simulation (not shared stores)

### Mock Theater (tests wiring, not behavior)
These test that wires were connected, not that the connected system works. They should be rewritten using the MemoryTokenStore pattern from `BucketFailoverHandlerImpl.spec.ts`:
- **`oauth-manager.failover-wiring.spec.ts`** — vi.spyOn on private `getProfileBuckets`, verifies `setBucketFailoverHandler` was called but not that failover actually works
- **`oauth-manager.getToken-bucket-peek.spec.ts`** — vi.mock on entire modules (`runtimeSettings`, `@vybestack/llxprt-code-core`), weakens MB-11 coverage

---

## Critical Gaps (Zero Coverage)

| Gap | Severity |
|-----|----------|
| **SB-10**: Auth flow mid-turn timeout | 🔴 Critical |
| **UE-01–UE-08**: User entry point lifecycle triggers | 🔴 High |
| **SA-01–SA-04**: Subagent isolation scenarios | 🔴 High |
| **MB-14**: Cross-process `getOAuthToken` refresh per-bucket | 🟠 High |
| **SB-03/SB-04**: Complete `getToken` fallthrough chain (end-to-end) | 🟡 Medium |
| **MB-04**: `AllBucketsExhaustedError` class + `bucketFailureReasons` map | 🟡 Medium |
| **EC-01**: Network error (throws) vs null on refresh | 🟡 Medium |
| **RO-01–RO-03**: Multi-bucket RetryOrchestrator integration | 🟠 High |

---

## Implementation

- Each scenario above must have a corresponding behavioral test (total: ~58 scenarios)
- Tests must use real implementations where possible (no mock theater)
- Token store can use the file-based `KeyringTokenStore` with temp directories, or `MemoryTokenStore` for in-process tests
- Provider refresh uses controlled test doubles that simulate real behavior
- Cross-process scenarios simulate via shared `SecureStore`/lockDir (two `KeyringTokenStore` instances sharing the same backing store), not `child_process.fork`
- All tests in `packages/cli/src/auth/__tests__/behavioral/` with scenario ID prefix on each test description (e.g., `'SB-01: Valid token returns access_token'`)
- Rewrite `oauth-manager.failover-wiring.spec.ts` and `oauth-manager.getToken-bucket-peek.spec.ts` using the `MemoryTokenStore` pattern

### Priority Order for New Test Development
1. **SB-10** — user-visible failure mode, zero coverage
2. **UE-03, UE-04, UE-05** — turn boundary and mid-turn paths directly tied to agent reliability
3. **SA-01–SA-03** — subagent isolation regressions are silent and hard to debug
4. **RO-01–RO-03** — multi-bucket retry integration gap
5. **MB-14, EC-01** — cross-process and error-throw paths
6. **UE-01, UE-02, UE-06–UE-08, SA-04, EC-02–EC-04** — lifecycle and edge cases

## Tracking

- Parent: #1777
- Related: #1781, #1782
- Milestone: 0.10.0

### PR DESCRIPTION
**Pull request by acoliver**
## Summary

Implements the formal specification and behavioral test coverage required by #1783 for all OAuth bucket failover scenarios. Adds **58 behavioral tests** across 7 new test files organized by scenario category, plus rewrites 2 existing mock-theater test files to use real implementations.

Closes #1783

### New Behavioral Test Files (`packages/cli/src/auth/__tests__/behavioral/`)

| File | Scenarios | Tests |
|------|-----------|-------|
| `test-utils.ts` | Shared infrastructure | MemoryTokenStore, makeToken, factories |
| `single-bucket.behavioral.spec.ts` | SB-01..SB-10 | 10 |
| `multi-bucket.behavioral.spec.ts` | MB-01..MB-14, RO-01..RO-03 | 17 |
| `proactive-renewal.behavioral.spec.ts` | PR-01..PR-10 | 10 |
| `bridge-ui.behavioral.spec.ts` | BR-01..BR-05 | 5 |
| `user-entry-points.behavioral.spec.ts` | UE-01..UE-08 | 8 |
| `subagent-isolation.behavioral.spec.ts` | SA-01..SA-04 | 4 |
| `error-edge-cases.behavioral.spec.ts` | EC-01..EC-04 | 4 |

### Rewritten Mock-Theater Files (vi.mock eliminated)

| File | Tests |
|------|-------|
| `oauth-manager.getToken-bucket-peek.spec.ts` | 6 |
| `oauth-manager.failover-wiring.spec.ts` | 6 |

**Total: 70 tests (58 new behavioral + 12 rewritten)**

## Testing Approach

All behavioral tests use **real implementations** -- no `vi.mock()` for module replacement:

- **MemoryTokenStore**: In-memory TokenStore with bucket-scoped storage, real `listBuckets()` derivation
- **Real OAuthManager**: Constructor with real token coordination, bucket management, proactive renewal
- **Real BucketFailoverHandlerImpl**: 3-pass failover (peek/refresh/reauth) with classification
- **Real ProactiveRenewalManager**: Timer scheduling, backoff, external refresh detection
- **Real globalOAuthUI**: Singleton buffer/flush/delivery semantics
- **vi.spyOn()**: Used only for assertion verification and controlled behavior injection

## Scenario Coverage Map

- **Single-Bucket (SB)**: Valid token, expired+refresh, refresh failure, no refresh_token, 401/403 failover, cross-process refresh, proactive renewal, auth timeout
- **Multi-Bucket (MB)**: Primary valid, 429/401 failover, all-exhausted, refresh recovery, pass 3 reauth (success/fail), removed bucket, session override, peek loop, eager auth, concurrent coalescing, cross-process per-bucket
- **Retry Orchestrator (RO)**: 401 classify+switch, 429 classify+switch, resetSession clears triedBuckets
- **Proactive Renewal (PR)**: Schedule/skip/dedup, renewal success/fail/backoff, MAX failures stop, external refresh detection, proxy mode skip, profile reconfiguration
- **Bridge/UI (BR)**: Module-load bridge, pre-mount buffering, FIFO flush, cleanup/remount, direct delivery
- **User Entry Points (UE)**: Profile load/switch, turn reset, mid-turn renewal, 429 transparent failover, declined auth, all-tokens-die, state cleanup
- **Subagent Isolation (SA)**: Scoped session buckets, concurrent subagents, failover isolation, metadata-scoped eager auth
- **Error/Edge Cases (EC)**: Network throw vs null, lock stale detection, SIGINT interruption, logout+reauth lifecycle

## Verification

- All 848 auth tests pass (62 test files)
- Lint: 0 errors (warnings only -- pre-existing pattern)
- Typecheck: 0 errors in changed files (pre-existing StatsDisplay errors only)
- Format: clean
- Build: passes
- Smoke test: passes

### CODE REVIEW

**Comment by acoliver** [comment]
Addressed 6 of 8 CodeRabbit comments in commit 1bb310eb4:

**Fixed:**
1. Removed redundant refreshTokenResult option in EC-01
2. De-duplicated EC-02 describe/it names
3. Added proactiveManager null guard in afterEach
4. Replaced try-catch with .catch() in EC-03 and SB-10
5. Fixed listProviders() in MemoryTokenStore to return stored providers
6. Let real authenticateMultipleBuckets run in failover-wiring test 6

**Intentionally left as-is (scope decisions):**

- **SB-07 cross-process**: Uses a shared MemoryTokenStore to simulate a second process updating the store. Using two KeyringTokenStore instances with a shared backing store would require file-system setup and is outside the scope of this PR (the issue catalog specifies in-process simulation, not real cross-process).

- **UE-03 turn-boundary**: Tests handler.resetSession() directly because the turn-boundary wiring lives in useGeminiStream (a React hook in the UI layer). Testing through that entry point would require mounting React components, which is a different test category and scope.

File: packages/cli/src/auth/__tests__/behavioral/single-bucket.behavioral.spec.ts [hunk 1]
```diff
@@ -0,0 +1,305 @@
...
+    expect(result!.access_token).toBe('process-a-token');
+
+    const refreshedToken = makeToken('process-b-refreshed');
+    await tokenStore.saveToken(PROVIDER, refreshedToken);
+
+    const resultA = await managerA.getOAuthToken(PROVIDER);
+
+    expect(resultA).not.toBeNull();
+    expect(resultA!.access_token).toBe('process-b-refreshed');
+  });
```
[the remark this answers is not in the document]
Comment (acoliver) [reply packages/cli/src/auth/__tests__/behavioral/single-bucket.behavioral.spec.ts:194]: Intentional design choice: SB-07 uses a shared MemoryTokenStore to simulate cross-process behavior (two managers reading/writing the same store), matching the issue specification which says cross-process scenarios should simulate via shared SecureStore/lockDir, not child_process.fork. The MemoryTokenStore IS the shared backing store here. Using two KeyringTokenStore instances with temp directories is feasible but would add file-system setup complexity for the same behavioral coverage. Leaving as scope decision for reviewer.

File: packages/cli/src/auth/__tests__/behavioral/user-entry-points.behavioral.spec.ts [hunk 2]
```diff
@@ -0,0 +1,317 @@
...
+
+    const second = await handler.tryFailover({ triggeringStatus: 429 });
+    expect(second).toBe(false);
+
+    handler.resetSession();
+
+    const third = await handler.tryFailover({ triggeringStatus: 429 });
+    expect(third).toBe(true);
+    expect(handler.getCurrentBucket()).toBe('bucket-a');
+  });
```
[the remark this answers is not in the document]
Comment (acoliver) [reply packages/cli/src/auth/__tests__/behavioral/user-entry-points.behavioral.spec.ts:159]: Intentional scope boundary: UE-03 tests resetSession() at the handler level because the turn-boundary trigger (useGeminiStream.ts L73) is a React hook that requires a full UI render context. Testing through that entry point would require mounting React components and simulating user input, which is a different category of integration test. The behavioral test verifies the contract that resetSession() clears triedBuckets and allows fresh failover, which is what the scenario specification describes.
